Avoid querying `named_shape` to support `DShapedArray`s.

erick-xanadu opened the pull request:
Hi,

We are using JAX to compile Python to MLIR. We are currently exploring the implementation of dynamic tensor sizes support in JAX (i.e., `abstracted_axes`). I noticed that when using the option calling `make_jaxpr` with `abstracted_axes=something` and `return_shapes=True` an exception is raised. This is because `DShapedArray`s does not have a `named_shape` attribute which is needed in these lines of code:

```python
2447   @wraps(fun)
2448   @api_boundary
2449   def make_jaxpr_f(*args, **kwargs):
#      ... snip ...
2465     if return_shape:
2466       out_avals, _ = unzip2(out_type)
2467       out_shapes_flat = [
2468           ShapeDtypeStruct(a.shape, a.dtype, a.named_shape) for a in out_avals] # <-- Here
2469       return closed_jaxpr, tree_unflatten(out_tree(), out_shapes_flat)
2470     return closed_jaxpr
```

I am not too familiar with Dynamic shapes in JAX yet, but implementing the changes allowed me to return a meaningful shape with the following value:

```
ShapeDtypeStruct(shape=(InDBIdx(val=0),), dtype=int64)
```

which appears to be sufficient.

Happy to hear your thoughts here. I would really like to be able to use `abstracted_axes` and `return_shapes=True`.

Thank you! :)

mattjj commented:
Thanks for the suggestion!

Actually, `named_shape` is kind of dead at the moment, so instead of adding it to a new place, I'd rather edit any place that expects the `named_shape` attribute so as not to expect it. Does that make sense?

erick-xanadu committed:
Don't assume shaped_name field exists in out_avals

`return_shape` and `abstracted_axes` arguments were not able to be used
at the same time when using `make_jaxpr`. Doing so will result in a
TypeError as the `out_avals` are expected to have a `shaped_name` field.
`DShapedArray`s do not contain `shaped_name`s.

erick-xanadu commented:
Hi @mattjj,

Thanks for your feedback. There are several ways to fix this, I took the least invasive approach here, just modifying the `make_jaxpr_f` function to not assume that `out_avals` have the `shaped_name` attribute. Would you like me to go through all the code and change `${var}.named_shape` to a function that checks if `named_shape` is present or not?

For the time being, the only change that is impeding our project is this line (and another PR I'll open to address the `get_aval` #18561 from the caller of `get_aval`).

Thanks!

erick-xanadu commented:
I'm closing as this has already been implemented.
